---
title: "16S Phylogenetic Tree Building"
Author: "Sarah Lucas"
---
# Setup environment
```{r}
.cran_packages <- c("tidyverse", "ips")
.bioc_packages <- c("dada2", "DECIPHER", "phangorn", "ShortRead")

.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
   BiocManager::install(.bioc_packages[!.inst], quietly = FALSE)
}
# Load packages into session, and print package version
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)

# Setting seed for reproducibility
set.seed(12345)
```

# Import Data
```{r}
seqtab <- readRDS("16S_output/seqtab.rds")
```

# Construct phylogenetic tree using the Phangorn R package (method suggested by Callahan/Holmes https://f1000research.com/articles/5-1492/v2)
```{r}
# seqtab is the sample:ASV table made in DADA2 - it should contain all samples and ASVs
seqs <- getSequences(seqtab)
names(seqs) <- seqs # This propogates the tip labels of the tree
alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA)

phang.align <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phang.align)
treeNJ <- NJ(dm)
fit = pml(treeNJ, data=phang.align)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                      rearrangement = "stochastic", control = pml.control(trace = 0))
detach("package:phangorn", unload=TRUE)
saveRDS(fitGTR, "16S_output/fitGTR.rds")
```