---
title: "FESS MMM Enrichment Phyloseq"
author: "Sarah Lucas"
output: 
  html_document:
    theme: cosmo
    toc: yes
  pdf_document: default
---

# Set up
```{r}
.cran_packages <- c("tidyverse","gridExtra", "ggthemes",
                    "ggpubr", "naniar", "scales","ampvis2")
.bioc_packages <- c("phyloseq","Biobase")

# Install CRAN packages (if not already installed)
.inst <- .cran_packages %in% installed.packages()
if (any(!.inst)){
  install.packages(.cran_packages[!.inst],repos = "http://cran.rstudio.com/")
}

.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
   BiocManager::install(.bioc_packages[!.inst], quietly = FALSE)
}

# Load packages into session, and print package version
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)
sapply(c(.cran_packages, .bioc_packages), package.version)
```

## Plotting aesthetics

Make figures transparent
```{r}
theme_transparent <- theme(panel.background = element_rect(fill = "transparent"),
                           plot.background = element_rect(fill = "transparent", 
                                                        color = NA),
                           panel.grid.major = element_line(color = "transparent"),
                           panel.grid.minor = element_line(color = "transparent"),
                           legend.background = element_rect(fill = "transparent"), 
                           legend.box.background = element_rect(fill = "transparent", 
                                                                color = "transparent"))
```



# Load in data
Load in all components created in the DADA2 analysis and the sample sheet
```{r}
samdf <- read.csv("../data/Enrichment_samdf.csv")
samdf <- data.frame(samdf)
rownames(samdf) <- samdf$SAMPLE_ID #because R - dplyr will not like making rownames, but it's fine for now.
# seqtab is the sample:ASV table made in DADA2 - it should contain all samples and ASVs
seqtab <- readRDS("../data/DADA2/seqtab_MAR2020.rds")
# Table with all ASVs and assigned taxonomy using the SILVA database (Greengenes and RDP also available)
taxtab <- readRDS("../data/DADA2/tax_species_final_silva_HOMD_MAR2020.rds")
# Phylogenetic tree made using DECIPHER and Phangorn - import object and pull tree from the fitGTR
fitGTR <- readRDS("../data/DADA2/fitGTR_MAR2020.rds")
```

## Make a phyloseq object
```{r}
library(phyloseq)
ps <- phyloseq(otu_table(seqtab, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxtab),
               phy_tree(fitGTR$tree)
)
ps
```
Sample Names:
```{r}
sample_names(ps)
```
Additional taxonomy levels for plotting purposes
```{r}
tax.PHYSEQ <-  data.frame(tax_table(ps)) %>%
  rownames_to_column('ASV') %>%
  #Now getting into formatting the taxa so it looks nice and informative in plots
  mutate(Genus_Species = ifelse(!is.na(Species), 
                                str_c(Genus, Species, sep = " "), as.character(Genus))) %>%
  mutate(Genus_Family = ifelse(!is.na(Genus), as.character(Genus), as.character(Family))) %>%
  select(ASV, Kingdom, Phylum, Class, Order, Family, Genus, Species, Genus_Family, Genus_Species, DB)
# Get rid of any extra notation in the taxonomy names.
tax.PHYSEQ <- data.frame(lapply(tax.PHYSEQ, function(x) {
  gsub("_[0-9]", "", x)
}))

# Make it a matrix again
tax.PHYSEQ <- tax.PHYSEQ %>%
  column_to_rownames('ASV') %>%
  as.matrix()
tax_table(ps) <- tax.PHYSEQ
ps

```


```{r}
# Remove samples that do not have further characterization experiments associated with them
removeSamples <- c("231","238","242","232","247")
ps <- subset_samples(ps, !SAMPLE_TYPE == "control" &
                             !PATIENT_NUM %in% removeSamples
                             )
ps
```
Remove any seqeunces that don't belong to any samples and/or have no counts
```{r}
MINREADS <- 0
ps <- filter_taxa(ps, function(x) sum(x) > MINREADS, TRUE)
ps
```
Read Summary
```{r}
summary(sample_sums(ps))
```

# Data Pre-processing

## Supervised Prevalence Filtering - Filter out any taxa that have an unassigned phylum - those aren't really useful to us.

```{r}
#Only want Bacterial Seqs, no chloroplast DNA, no mitochondrial DNA
filterKingdom = c("Archaea", "Eukaryota")
filterOrder = "Chloroplast" ##Chloroplast DNA
filterFamily = "Mitochondria" ## Mitochonidrial DNA

#filter features with an ambiguous phylum annotation.
ps.pv.1 <- subset_taxa(
  ps,!is.na(Phylum) &
    !Phylum %in% c("", "uncharacterized", "Unknown_Phylum") &
    !Kingdom %in% filterKingdom &
    !Order %in% filterOrder &
    !Family %in% filterFamily
)
ps.pv.1
```
# Alpha diversity
Calculate alpha diversity here while 
```{r}

```





A useful next step is to explore feature prevalence in the dataset, which we will define here as the number of samples in which a taxa appears at least once.
```{r}
# Compute prevalence of each feature, store as data.frame
prevdf = apply(
  X = otu_table(ps.pv.1),
  MARGIN = ifelse(taxa_are_rows(ps.pv.1), yes = 1, no = 2),
  FUN = function(x) {
    sum(x > 0)
  }
)
# Add taxonomy and total read counts to this data.frame
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps.pv.1),
                    tax_table(ps.pv.1))
```
Take a look graphically at the distribution of taxa among the different phyla:
```{r}
# Subset to the remaining phyla after initial filtering
prevdf.pv.phylum <- subset(prevdf, Phylum %in% get_taxa_unique(ps.pv.1, "Phylum"))
ggplot(prevdf.pv.phylum,
       aes(TotalAbundance, Prevalence / nsamples(ps.pv.1), color = Phylum)) +
  # Include a guess for filtering parameter at 0.02
  geom_hline(yintercept = 0.02,
             alpha = 0.5,
             linetype = 2) + 
  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  theme(text = element_text(size = 10)) +
  facet_wrap( ~ Phylum) + theme(legend.position = "none")
```
## Are there phyla that are comprised of mostly low-prevalence features? Compute the total and average prevalences of the features in each phylum:
```{r}
#just look at phylum
prevdf.phylum <- plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
colnames(prevdf.phylum)[2] <- "mean_feature_prevalence"
colnames(prevdf.phylum)[3] <- "total_feature_prevalence"
prevdf.phylum
```
Yes. Looks like there are some phyla that have quite low prevalence. Let's filter out any fatures belonging to phyla that have an average prevalence of just 1, or have less than 10 features:
```{r}
prevdf.phylum.filt <- filter(prevdf.phylum, mean_feature_prevalence <= 1 | total_feature_prevalence < 10)
prevdf.phylum.filt
```
Remove the features belonging to these phyla in prevdf.phylum.filt from the dataset, creating ps.pv.2
```{r}
# Filter low prevalence samples (supervised - you made this decision by looking at your data)
# Define phyla to filter - Here I have chosen those phyla that have an average prevalence of 1 OR have a total number of features less than 10.
filterPhyla = unique(prevdf.phylum.filt$Phylum)
filterPhyla
# Filter entries with unidentified Phylum.
ps.pv.2 <- subset_taxa(ps.pv.1, !Phylum %in% filterPhyla)
ps.pv.2
#tax.CRS.pv.2 <- data.frame(tax_table(CRS.pv.2))
```

# Unsupervised Prevalence Filtering

The previous filtering steps are considered supervised, because they relied on prior information that is external to this experiment (a taxonomic reference database). This next filtering step is completely unsupervised, relying only on the data in this experiment, and a parameter that we will choose after exploring the data. Thus, this filtering step can be applied even in settings where taxonomic annotation is unavailable or unreliable.

First, explore the relationship of prevalence and total read count for each feature. Sometimes this reveals outliers that should probably be removed, and also provides insight into the ranges of either feature that might be useful. This aspect depends quite a lot on the experimental design and goals of the downstream inference, so keep these in mind. It may even be the case that different types of downstream inference require different choices here. There is no reason to expect ahead of time that a single filtering workflow is appropriate for all analyses.

Next, combine those ASVs that have the same TaxName. This will combine phylogenetically similar sequences at distinct levels of similarity.

```{r}
#  Define prevalence threshold as % of total samples
prevalenceThreshold = 1
prevalenceThreshold

# Execute prevalence filter, using `prune_taxa()` function
keepTaxa <-  rownames(prevdf)[(prevdf$Prevalence >= prevalenceThreshold)]
ps.pv.unsuper <- prune_taxa(keepTaxa, ps.pv.2)
ps.pv.unsuper
```
No difference there, as expected. All taxa in the dataset should belong to at least one sample.

abundance filtering:
```{r}
ps.pv.unsuper.prevfilt.prop <- transform_sample_counts(ps.pv.unsuper, function(x){x / sum(x)})
ps.pv.unsuper.prevfilt.prop
```

Filter out taxa with less than .0001 relative abundance
```{r}
abundanceThreshold <- .0001
ps.pv.unsuper.prevfilt.taxfilt.prop = filter_taxa(ps.pv.unsuper.prevfilt.prop, function(x) mean(x) > abundanceThreshold, TRUE)
ps.pv.unsuper.prevfilt.taxfilt.prop
keepTaxaRelAbund <- taxa_names(ps.pv.unsuper.prevfilt.taxfilt.prop)

#keepTaxaAbund <-  rownames(prevdf3)[(prevdf3$TotalAbundance >= abundanceThreshold)]
ps.pv.unsuper.prevfilt.taxfilt <- prune_taxa(keepTaxaRelAbund, ps.pv.unsuper)
ps.pv.unsuper.prevfilt.taxfilt
```

```{r}
summary(sample_sums(ps.pv.unsuper.prevfilt.taxfilt))
```
```{r}
which(!rowSums(otu_table(ps.pv.unsuper.prevfilt.taxfilt)) > 2000)
seqThreshold <- 2000
#ps.pv.unsuper.prevfilt.taxfilt2000 <- prune_samples(sample_sums(ps.pv.unsuper.prevfilt.taxfilt) > seqThreshold, ps.pv.unsuper.prevfilt.taxfilt)
#ps.pv.unsuper.prevfilt.taxfilt2000
```

Final prevalence/abundance analysis
```{r}
# Compute prevalence of each feature, store as data.frame
prevdf.filt = apply(X = otu_table(ps.pv.unsuper.prevfilt.taxfilt),
                    MARGIN = ifelse(taxa_are_rows(ps.pv.unsuper.prevfilt.taxfilt), yes = 1, no = 2),
                    FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevdf.filt = data.frame(Prevalence = prevdf.filt,
                         TotalAbundance = taxa_sums(ps.pv.unsuper.prevfilt.taxfilt),
                         tax_table(ps.pv.unsuper.prevfilt.taxfilt))
```

Take a look graphically at the distribution of taxa among the different phyla:
```{r}
# Subset to the remaining phyla after initial filtering
prevdf.filt.phylum = subset(prevdf.filt, Phylum %in% get_taxa_unique(ps.pv.unsuper.prevfilt.taxfilt, "Phylum"))
ggplot(prevdf.filt.phylum, aes(TotalAbundance, Prevalence / nsamples(ps.pv.unsuper.prevfilt.taxfilt),color=Phylum)) +
  # Include a guess for filtering parameter at 0.02
  geom_hline(yintercept = 0.02, alpha = 0.5, linetype = 2) + geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  theme(text = element_text(size=10)) +
  facet_wrap(~Phylum) + theme(legend.position="none")
```


# Abundance Analysis

## Top Genera
```{r}
original <- subset_samples(ps.pv.unsuper.prevfilt.taxfilt, SAMPLE_TYPE == "original") 
enrich <- subset_samples(ps.pv.unsuper.prevfilt.taxfilt, SAMPLE_TYPE == "enrichment") 
```

What are the taxa making up most of the reads in each group - the original samples and the enrichment samples
```{r}
#what are the abundance levels of each genus?
original.genus <- tax_glom(original, taxrank="Genus")
tax.count.ori <- data.frame(tax_table(original.genus)[,2:6], taxa_sums(original.genus))
rownames(tax.count.ori) = NULL
colnames(tax.count.ori) <- c("Phylum","Class","Order","Family","Genus", "Abundance")
tax.count.ori$Percent <- round(tax.count.ori$Abundance/sum(tax.count.ori$Abundance)*100, 4)
Genus_df_ori <- tax.count.ori[with(tax.count.ori, order(-Percent)), ] 

#how much do the top 10 genera contribute to total abundance?
top10GeneraOri <- Genus_df_ori[1:10, ]
round(sum(top10GeneraOri$Percent),3)
```
```{r}
###How diverse are the top 10 genera? i.e., how many species are there per genus?
top10GeneraOri <- as.vector(Genus_df_ori$Genus[1:10])
Diversity.list.ori <- vector("list", 10)
names(Diversity.list.ori) <- top10GeneraOri

for (i in 1:length(top10GeneraOri)) {
  physub = subset_taxa(original, Genus == top10GeneraOri[i])
  physub = prune_taxa(taxa_sums(physub) > 0, physub)
  Diversity.list.ori[[i]] <- physub
}

#compute the number of taxa in each element of the list
NtaxaOri <- data.frame(unlist(lapply(Diversity.list.ori, ntaxa)))

colnames(NtaxaOri) <- "N.Species"
#Make a table with percent abundance and number of taxa
genus.tab.ori <- data.frame(Genus_df_ori[1:10, ], NtaxaOri)
genus.tab.ori
```

```{r}
#what are the abundance levels of each genus?
enrichment.genus <- tax_glom(enrich, taxrank = "Genus")
tax.count.enrich <-
  data.frame(tax_table(enrichment.genus)[, 2:6], taxa_sums(enrichment.genus))
rownames(tax.count.enrich) = NULL
colnames(tax.count.enrich) <-
  c("Phylum", "Class", "Order", "Family", "Genus", "Abundance")
tax.count.enrich$Percent <-
  round(tax.count.enrich$Abundance / sum(tax.count.enrich$Abundance) * 100,
        4)
Genus_df_enrich <-
  tax.count.enrich[with(tax.count.enrich, order(-Percent)),]

#how much do the top 10 genera contribute to total abundance?
top10GeneraEnrich <- Genus_df_enrich[1:10,]
round(sum(top10GeneraEnrich$Percent), 3)
```
```{r}
###How diverse are the top 10 genera? i.e., how many species are there per genus?
top10GeneraEnrich <- as.vector(Genus_df_enrich$Genus[1:10])
Diversity.list.enrich <- vector("list", 10)
names(Diversity.list.enrich) <- top10GeneraEnrich

for (i in 1:length(top10GeneraEnrich)) {
  physub = subset_taxa(enrich, Genus == top10GeneraEnrich[i])
  physub = prune_taxa(taxa_sums(physub) > 0, physub)
  Diversity.list.enrich[[i]] <- physub
}

#compute the number of taxa in each element of the list
NtaxaEnrich <-
  data.frame(unlist(lapply(Diversity.list.enrich, ntaxa)))

colnames(NtaxaEnrich) <- "N.Species"
#Make a table with percent abundance and number of taxa
genus.tab.enrich <- data.frame(Genus_df_enrich[1:10, ], NtaxaEnrich)
genus.tab.enrich
```

## Taxa Plots

Transform to proportions
```{r}
ps.pv.unsuper.prevfilt.taxfilt.prop <- transform_sample_counts(ps.pv.unsuper.prevfilt.taxfilt, function(x){100 * x / sum(x)})
ps.pv.unsuper.prevfilt.taxfilt.log <- transform_sample_counts(ps.pv.unsuper.prevfilt.taxfilt, function(x) log(1 + x))
```

Melt phyloseq object:
To Do: Add on extra taxa to spreadsheet with colors.
```{r}
color_hex <- read_csv("../ColorHex.csv")
```

```{r}
ps.melt <- psmelt(ps.pv.unsuper.prevfilt.taxfilt.prop) %>%
  select(SAMPLE_TYPE, SAMPLE_ID, PATIENT_NUM, PUB_LETTER,
         Kingdom, Phylum, Class, Order, Family, Genus, Species, Genus_Species, Genus_Family, Abundance) %>%
  mutate_at(vars(Kingdom:Genus_Family),funs(as.character)) %>%
  mutate(GenusFamilyPlotting = ifelse(Abundance > 1, Genus_Family, "Other < 1%")) %>%
  left_join(color_hex, by = c("GenusFamilyPlotting" = "Taxon")) %>%
  mutate(PhylumOther = ifelse(GenusFamilyPlotting == "Other < 1%", "zOther", Phylum.x)) %>%
  arrange(PhylumOther, GenusFamilyPlotting, Abundance)
# Get unique names for plotting
unique(ps.melt$GenusFamilyPlotting)
```

```{r}
colorTax <- unique(ps.melt$GenusFamilyPlotting)
colorHex <- unique(ps.melt$Color)
ps.melt$GenusFamilyPlotting <- factor(ps.melt$GenusFamilyPlotting, levels=colorTax)

#fig.height = 4, fig.width = 7.5
## How many unique names are we dealing with (for plotting)
barplot_colors <- ggplot(data=ps.melt, aes(x = as.character(SAMPLE_ID), 
                                           y = Abundance, 
                                           fill = GenusFamilyPlotting)) +
  facet_grid(~PUB_LETTER, scales = "free", space = "free") +
  geom_bar(aes(), stat="identity", position="stack") + 
#  scale_fill_manual(breaks = colorHex, 
#                    values = colorTax) +
  scale_x_discrete(labels = c("O","E"))+
  #theme_pubr(base_size = 10) + 
  theme(legend.position = "right", 
        legend.title = element_blank(),
        legend.text = element_text(face = "italic"),
        axis.ticks.y = element_line(color = "black"),
        axis.ticks.x = element_blank(),
#        strip.text.x = element_blank(),
#        strip.background = element_blank()
  ) + theme_transparent +
  labs(x = "", y = "Relative Abundance (% reads)") #+
  #guides(fill = guide_legend(ncol=2, override.aes = list(size = 5)))
barplot_colors
```
```{r}
#ggsave(barplot_colors, filename = "~/Documents/MICaB/Hunter_Lab/Projects/inProgress/CRS_MMM_Enrichments/CRS_Enrichments_16S_Sequencing_July2018/figures/phyloseq/genus_family_taxa_table_filt002_legendright.png", height = 4, width = 8, bg = "transparent")
#ggsave(barplot_colors, filename = "~/Documents/MICaB/Hunter_Lab/Projects/inProgress/CRS_MMM_Enrichments/CRS_Enrichments_16S_Sequencing_July2018/figures/phyloseq/genus_family_taxa_table_filt002_legendright.pdf", height = 4, width = 8, bg = "transparent", useDingbats = FALSE)
```

## Merged Stacked Bar Plots
```{r}
ps.pv.unsuper.prevfilt.taxfilt.genusFamily <- tax_glom(ps.pv.unsuper.prevfilt.taxfilt, "Genus_Species")
merged <- merge_samples(ps.pv.unsuper.prevfilt.taxfilt.genusFamily, "SAMPLE_TYPE")
mergedProp <- transform_sample_counts(merged, function(x){100 * x / sum(x)})
mergedPropMelt <- psmelt(mergedProp) %>%
  mutate(Genus_Species = ifelse(Abundance > 1, as.character(Genus_Species), "Other < 1%"))
unique(mergedPropMelt$Genus_Species)
```

```{r}
color_hex <- read_csv("../ColorHex.csv")
mergedPropMelt <- mergedPropMelt %>% left_join(color_hex, by = c("Genus_Species" = "Taxon")) %>%
  mutate(PhylumOther = ifelse(Genus_Species == "Other < 1%", "zOther", Phylum.x)) %>%
  arrange(PhylumOther, Genus_Species, Abundance)
```

```{r fig.height=6.5, fig.width=5.5}
colorTax <- unique(mergedPropMelt$Genus_Species)
colorHex <- unique(mergedPropMelt$Color)
mergedPropMelt$Genus_Species <- factor(mergedPropMelt$Genus_Species, levels=colorTax)

mergedPropMeltPlot <- ggplot(mergedPropMelt, aes(x = Sample, y = Abundance, fill = Genus_Species)) +
  labs(x = "", y = "Relative Abundance (% reads)") +
  geom_bar(aes(), stat="identity", position="stack") + 
  scale_fill_manual(breaks = colorTax,
                    values = colorHex) +
  scale_x_discrete(
    breaks = c("original","enrichment"),
    limits = c("original", "enrichment"),
    labels = c("Original","Enrichment")) +
  theme_pubr(base_size = 10) + 
  theme(legend.position = "right", 
        legend.title = element_blank(),
        axis.ticks.y = element_line(color = "black"),
        axis.text.x = element_text(angle = 45, hjust=1, color = "black"),
        axis.ticks.x = element_blank(),
        legend.text = element_text(face = "italic",size = 8, color = "black"),
        legend.key.size = unit(.35, "cm")) +
  guides(fill = guide_legend(ncol = 1))
mergedPropMeltPlot
```
```{r}
ggsave(mergedPropMeltPlot, filename = "../figures/enrichment_boxplot_grouped.pdf",
       device = "pdf",
       height = 4.5,
       width = 4.5)
```
## Heatmap
```{r}
library("ampvis2")
#Combine OTU abundance table and taxonomy table from the phyloseq object "my_phyloseq_object":
obj <- ps.pv.unsuper
# Make short names for OTUs - facilitates later plotting ease when amp_heatmap has the option tax_empty = "best" (for some reason)
taxa_names(obj) <- paste0("ASV", seq(ntaxa(obj)))
# Fix OTU table layout for exporting. taxa_as_rows = FALSE was not working.
Totu_table =t(otu_table(obj))
otu_table(obj)=Totu_table
#export OTU table from phyloseq object
otutable <- data.frame(OTU = rownames(phyloseq::otu_table(obj)@.Data),
                       phyloseq::otu_table(obj)@.Data,
                       phyloseq::tax_table(obj)@.Data,
                       check.names = FALSE
                       )
otutable <- otutable %>% select(-OTU,-Genus_Species, -Genus_Family, -DB)

#Extract metadata from the phyloseq object:
metadata <- data.frame(phyloseq::sample_data(obj), 
                       check.names = FALSE
                       )
rownames(metadata) <- c()
metadata$SAMPLE_TYPE <- factor(metadata$SAMPLE_TYPE, levels=c("original","enrichment"))
metadata <- mutate(metadata, SAMPLE_TYPE_CAPS = ifelse(SAMPLE_TYPE == "original", "Original","Enrichment"))
metadata$SAMPLE_TYPE_CAPS <- factor(metadata$SAMPLE_TYPE_CAPS, levels=c("Original","Enrichment"))


# Extract phylogenetic tree from phyloseq object:
phytree <- phyloseq::phy_tree(obj)

#Load the data with amp_load:
ampvis.obj <- amp_load(otutable, metadata,
                        tree = phytree
                        )
ampvis.obj
```

```{r, fig.height = 5.5, fig.width = 7.5}
heatmap <- amp_heatmap(data = ampvis.obj,
                       group_by = "SAMPLE_TYPE",
                       facet_by = "PUB_LETTER",
                       tax_aggregate = "Genus",
                       #tax_add = "Genus",
                       tax_empty = "Family",
                       plot_values = FALSE,
                       tax_show = 20,
                       plot_colorscale = "log10",
                       plot_na = TRUE,
                       normalise = TRUE,
                       plot_legendbreaks = c(.1,1,10),
                       color_vector = c("#f2f9fc", "#0072B2")
                      ) +
  theme(axis.text.y = element_text(face = "italic", size = 11),
        axis.text.x = element_text(size = 11),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10))
heatmap
```

```{r}
#ggsave(heatmap, filename = "../figures/enrichment_heatmap.pdf",
#       device = "pdf",
#       height = 5.5,
#       width = 6.6)
```

```{r, fig.height=5.5, fig.width = 5}
heatmapGrouped <- amp_heatmap(data = ampvis.obj,
                       group_by = "SAMPLE_TYPE_CAPS",
                       facet_by = "SAMPLE_TYPE_CAPS",
                       tax_aggregate = "Genus",
                       #tax_add = "Genus",
                       tax_empty = "Family",
                       plot_values = TRUE,
                       tax_show = 20,
                       plot_legendbreaks = c(.1,1,10,100),
                       plot_colorscale = "log10",
                       plot_na = TRUE,
                       normalise = TRUE,
                       color_vector = c("#f2f9fc", "#0072B2")
                      ) +
  theme(axis.text.y = element_text(face = "italic", size = 11, color = "black"),
        axis.text.x = element_blank(),
        axis.ticks  = element_blank(),
        legend.position = "right",
        strip.text = element_text(size = 11, color = "black"),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10),
        )
heatmapGrouped
```
```{r}
ggsave(heatmapGrouped, filename = "../figures/enrichment_heatmap_grouped.pdf",
       device = "pdf",
       height = 5.5,
       width = 5)
```

# Ordination
## Double Principal Coordinate Analysis (DPCoA)
Transformations
```{r}
ps.prop <- transform_sample_counts(ps.pv.unsuper.prevfilt.taxfilt, function(x){100 * x / sum(x)})
ps.log <- transform_sample_counts(ps.pv.unsuper.prevfilt.taxfilt, function(x) log(1 + x))
```

```{r}
# Calculate dpcoa 
ps.dpcoa <- ordinate(ps.prop, method = "DPCoA")
dpcoa_evals <- ps.dpcoa$eig
# Plot
ps.dpcoa.plot = plot_ordination(ps.prop, ps.dpcoa, type="biplot", color="Phylum", shape="SAMPLE_TYPE") +
  geom_point(alpha = 1, size = 3) +
  theme_pubr(base_size = 10) +
  theme(
    legend.position = "right",
    legend.text = element_text(color = "black", face = "bold"),
    legend.title = element_blank(),
    axis.line = element_line(color = "black"),
    axis.ticks = element_line(color = "black"),
    axis.text = element_text(color = "black"),
    axis.title = element_text(color = "black")
    ) +
 # guides(color = guide_legend(nrow = 1, override.aes = list(fill="transparent"))) +
  coord_fixed(sqrt(dpcoa_evals[2] / dpcoa_evals[1])) +
  scale_color_manual(values = c("black","#de2d26", "#009E73","#56B4E9","#E69F00","#F0E442","#D55E00","#0072B2")) +
  scale_shape_manual(values = c(4,13,16)) #For catagorical variables

#Get rid of overlapping layers of points
ps.dpcoa.plot$layers <- ps.dpcoa.plot$layers[-1]
ps.dpcoa.plot
```

```{r}
ggsave(ps.dpcoa.plot, filename = "../figures/dpcoa_biplot.pdf",
       device = "pdf",
       useDingbats = FALSE,
       height = 5,
       width = 6)
```




## Principal Coordinate Analysis (PCoA) - Weighted Unifrac
```{r}
ps.wunifrac.ord <- ordinate(ps.prop, distance = "wunifrac", "PCoA")
```
```{r}
ps.wunifrac.ord.plot <- plot_ordination(ps.prop, ps.wunifrac.ord, type="split", color="Phylum", shape="SAMPLE_TYPE") +
  geom_point(alpha = 1, size = 3) +
  theme_pubr(base_size = 10) +
  theme(
    legend.position = "right",
    legend.text = element_text(color = "black", face = "bold"),
    legend.title = element_blank(),
    axis.line = element_line(color = "black"),
    axis.ticks = element_line(color = "black"),
    axis.text = element_text(color = "black"),
    axis.title = element_text(color = "black")
    ) +
 # guides(color = guide_legend(nrow = 1, override.aes = list(fill="transparent"))) +
  coord_fixed(sqrt(dpcoa_evals[2] / dpcoa_evals[1])) +
  scale_color_manual(values = c("black","red", "#009E73","#56B4E9","#E69F00","#F0E442","#D55E00","#0072B2")) +
  scale_shape_manual(values = c(20,10,16)) #For catagorical variables

#Get rid of overlapping layers of points
ps.wunifrac.ord.plot$layers <- ps.wunifrac.ord.plot$layers[-1]
ps.wunifrac.ord.plot
```


# Taxonomy: SILVA vs HOMD
What were the taxa in the filtered dataset that were contributed by the HOMD database?
```{r}
taxFinal <- data.frame(tax_table(ps.pv.unsuper.prevfilt.taxfilt))
dim(taxFinal)

HOMDtax <- data.frame(tax_table(ps.pv.unsuper.prevfilt.taxfilt)) %>%
  filter(DB == "HOMD")
dim(HOMDtax)

NonHOMDtax <- data.frame(tax_table(ps.pv.unsuper.prevfilt.taxfilt)) %>%
  filter(!DB == "HOMD" | is.na(DB))
dim(NonHOMDtax)
```

```{r}
HOMDNotInSILVA <- anti_join(HOMDtax, NonHOMDtax, by = "Genus_Species")
dim(HOMDNotInSILVA)
```
47 of the 50 HOMD IDd taxa were uniquely added by HOMD.
```{r}
HOMDInSILVA <- anti_join(HOMDtax, HOMDNotInSILVA, by = "Genus_Species")
dim(HOMDInSILVA)
```

